File: python/src/linodemcp/twostage/store.py
"""In-memory two-stage plan store.

Mirrors ``go/internal/twostage/store.go``. Plans live only in process
memory; a restart drops them all. A janitor task reaps expired plans on an
interval, and ``put`` enforces a hard ceiling by evicting the oldest plan.
Concurrency is guarded by an ``asyncio.Lock`` since the server dispatches
tool calls on the event loop.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable
    from datetime import timedelta

    from mcp.types import TextContent

# Once the store holds this many plans, the next put evicts the oldest by
# planned_at to make room. A later apply against an evicted plan looks the same
# as an unknown id.
MAX_OUTSTANDING_PLANS = 1000


class PlanNotFoundError(Exception):
    """Raised when no plan with the given id exists (or it was consumed)."""


class PlanExpiredError(Exception):
    """Raised when a plan exists but its TTL has elapsed."""


@dataclass
class PlanEntry:
    """One outstanding plan held in process memory."""

    id: str
    tool: str
    environment: str
    args: dict[str, Any]
    state_hash: str
    planned_at: datetime
    expires_at: datetime
    apply: Callable[[], Awaitable[list[TextContent]]]
    # Normalized top-level field map of the planned state (hash-ignore fields
    # already stripped). On a drift refusal the apply path diffs it against the
    # re-fetched state to name the changed fields. None when the state did not
    # serialize to a JSON object. Mirrors the Go PlanEntry.StateFields.
    state_fields: dict[str, Any] | None = None


def _wall_clock() -> datetime:
    return datetime.now(UTC)


def _empty_plans() -> dict[str, PlanEntry]:
    return {}


@dataclass
class PlanStore:
    """Holds outstanding plans, guarded by an asyncio lock."""

    now: Callable[[], datetime] = _wall_clock
    _plans: dict[str, PlanEntry] = field(default_factory=_empty_plans)
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock)

    async def put(self, entry: PlanEntry) -> None:
        """Store a plan, evicting the oldest entry first when at the ceiling."""
        async with self._lock:
            if len(self._plans) >= MAX_OUTSTANDING_PLANS:
                self._evict_oldest_locked()
            self._plans[entry.id] = entry

    async def get(self, plan_id: str) -> PlanEntry:
        """Return a plan without consuming it.

        Raises PlanNotFoundError for an unknown id and PlanExpiredError when
        the TTL has elapsed.
        """
        async with self._lock:
            entry = self._plans.get(plan_id)
            if entry is None:
                raise PlanNotFoundError(plan_id)
            if self.now() > entry.expires_at:
                raise PlanExpiredError(plan_id)
            return entry

    async def take(self, plan_id: str) -> PlanEntry:
        """Return a plan and remove it atomically (single-use)."""
        async with self._lock:
            entry = self._plans.pop(plan_id, None)
            if entry is None:
                raise PlanNotFoundError(plan_id)
            if self.now() > entry.expires_at:
                raise PlanExpiredError(plan_id)
            return entry

    async def remove(self, plan_id: str) -> None:
        """Drop a plan by id. A no-op when the id is absent."""
        async with self._lock:
            self._plans.pop(plan_id, None)

    async def sweep(self) -> int:
        """Drop every expired plan, returning how many were removed."""
        async with self._lock:
            now = self.now()
            expired = [
                pid for pid, entry in self._plans.items() if now > entry.expires_at
            ]
            for pid in expired:
                del self._plans[pid]
            return len(expired)

    async def length(self) -> int:
        """Return the number of outstanding plans."""
        async with self._lock:
            return len(self._plans)

    def start_janitor(self, interval: timedelta) -> asyncio.Task[None]:
        """Launch a background task that sweeps expired plans on an interval."""

        async def _run() -> None:
            while True:
                await asyncio.sleep(interval.total_seconds())
                await self.sweep()

        return asyncio.create_task(_run())

    def _evict_oldest_locked(self) -> None:
        if not self._plans:
            return
        oldest = min(self._plans.values(), key=lambda entry: entry.planned_at)
        del self._plans[oldest.id]

File: go/internal/twostage/store.go
package twostage

import (
	"context"
	"sync"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
)

// MaxOutstandingPlans caps the store. Once it is reached, the next Put evicts
// the plan with the oldest PlannedAt to make room. A later apply against an
// evicted plan returns ErrPlanNotFound, the same as for an unknown ID.
const MaxOutstandingPlans = 1000

// ApplyFunc executes the planned action at apply time. It runs only after the
// drift check passes. It must resolve config live rather than capturing a
// plan-time snapshot, so a token rotation or environment change between plan
// and apply takes effect (see the spec's handler-closure rule).
type ApplyFunc func(ctx context.Context) (*mcp.CallToolResult, error)

// PlanEntry is one outstanding plan held in process memory.
type PlanEntry struct {
	Apply       ApplyFunc
	Args        map[string]any
	ID          string
	Tool        string
	Environment string
	StateHash   string
	// StateFields is the normalized top-level field map of the planned state
	// (hash-ignore fields already stripped). On a drift refusal the apply path
	// diffs it against the re-fetched state to report which fields changed. Nil
	// when the state did not serialize to a JSON object.
	StateFields map[string]any
	PlannedAt   time.Time
	ExpiresAt   time.Time
}

// PlanStore holds outstanding plans in process memory. Plans do not survive a
// restart. A janitor drops expired entries on an interval; Put also enforces a
// hard ceiling by evicting the oldest plan when the store is full.
type PlanStore struct {
	now   func() time.Time
	plans map[string]*PlanEntry
	mu    sync.RWMutex
}

// Option configures a PlanStore at construction time.
type Option func(*PlanStore)

// WithClock overrides the wall clock the store uses for expiry decisions.
// Tests inject a controllable clock; production passes nothing and gets
// time.Now.
func WithClock(now func() time.Time) Option {
	return func(s *PlanStore) {
		s.now = now
	}
}

// NewPlanStore returns an empty plan store. By default it reads the real wall
// clock; pass WithClock to override.
func NewPlanStore(opts ...Option) *PlanStore {
	store := &PlanStore{
		now:   time.Now,
		plans: make(map[string]*PlanEntry),
	}

	for _, opt := range opts {
		opt(store)
	}

	return store
}

// Put stores a plan, evicting the oldest entry first when the store is at its
// ceiling.
func (s *PlanStore) Put(entry *PlanEntry) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.plans) >= MaxOutstandingPlans {
		s.evictOldestLocked()
	}

	s.plans[entry.ID] = entry
}

// Get returns a plan without consuming it. It reports ErrPlanNotFound for an
// unknown ID and ErrPlanExpired when the TTL has elapsed.
func (s *PlanStore) Get(planID string) (*PlanEntry, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	entry, ok := s.plans[planID]
	if !ok {
		return nil, ErrPlanNotFound
	}

	if s.now().After(entry.ExpiresAt) {
		return nil, ErrPlanExpired
	}

	return entry, nil
}

// Take returns a plan and removes it in the same locked section, giving the
// apply path single-use semantics and preventing a concurrent double-apply.
// An expired plan is still removed; the caller receives ErrPlanExpired.
func (s *PlanStore) Take(planID string) (*PlanEntry, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	entry, ok := s.plans[planID]
	if !ok {
		return nil, ErrPlanNotFound
	}

	delete(s.plans, planID)

	if s.now().After(entry.ExpiresAt) {
		return nil, ErrPlanExpired
	}

	return entry, nil
}

// Remove drops a plan by ID. It is a no-op when the ID is absent.
func (s *PlanStore) Remove(planID string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.plans, planID)
}

// Sweep drops every expired plan and returns how many it removed. The janitor
// calls this on an interval; tests call it directly.
func (s *PlanStore) Sweep() int {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := s.now()

	var removed int

	for key, entry := range s.plans {
		if now.After(entry.ExpiresAt) {
			delete(s.plans, key)

			removed++
		}
	}

	return removed
}

// Len reports the number of outstanding plans.
func (s *PlanStore) Len() int {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return len(s.plans)
}

// StartJanitor runs a background loop that sweeps expired plans on the given
// interval until the context is canceled. It returns immediately; the loop
// owns its own goroutine.
func (s *PlanStore) StartJanitor(ctx context.Context, interval time.Duration) {
	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				s.Sweep()
			}
		}
	}()
}

// evictOldestLocked removes the plan with the earliest PlannedAt. The caller
// must hold the write lock.
func (s *PlanStore) evictOldestLocked() {
	var (
		oldestID string
		oldestAt time.Time
	)

	for key, entry := range s.plans {
		if oldestID == "" || entry.PlannedAt.Before(oldestAt) {
			oldestID = key
			oldestAt = entry.PlannedAt
		}
	}

	if oldestID != "" {
		delete(s.plans, oldestID)
	}
}
